managed/services/management/service.go

```diff
@@ -144,6 +146,96 @@ func (s *ServiceService) RemoveService(ctx context.Context, req *managementpb.Re
 	return &managementpb.RemoveServiceResponse{}, nil
 }
 
+// AddCustomLabels adds or replaces (if key exists) custom labels for a service.
+func (s *ServiceService) AddCustomLabels(ctx context.Context, req *managementpb.AddCustomLabelsRequest) (*managementpb.AddCustomLabelsResponse, error) {
+	if req.ServiceId == "" {
+		return nil, status.Error(codes.InvalidArgument, "service_id is required")
+	}
+
+	err := s.db.InTransactionContext(ctx, nil, func(tx *reform.TX) error {
+		service, err := models.FindServiceByID(tx.Querier, req.ServiceId)
+		if err != nil {
+			return err
+		}
+
+		labels, err := service.GetCustomLabels()
+		if err != nil {
+			return err
+		}
+		if labels == nil {
+			labels = make(map[string]string)
```

Comment:
Is it worth it to predefine capacity? how many labels are usually being added?

Reply:
Great catch. We could even do better and assign the labels directly, without iterating through the map.